File: 0119-pascals-triangle-ii/0119-pascals-triangle-ii.py
class Solution:
    def ncr(self, n, r, dp):
        if r == 0 or r == n:
            return 1
        if r == 1 or r == n - 1:
            return n
        if dp[n][r] != -1:
            return dp[n][r]
        dp[n][r] = self.ncr(n - 1, r, dp) + self.ncr(n - 1, r - 1, dp)
        return dp[n][r]

    def getRow(self, rowIndex):
        ans = []
        dp = [[-1] * (rowIndex + 1) for _ in range(rowIndex + 1)]
        for i in range(rowIndex + 1):
            ans.append(self.ncr(rowIndex, i, dp))
        return ans

File: 0119-pascals-triangle-ii/0119-pascals-triangle-ii.cpp
class Solution {
public:
    int ncr(int n,int r,vector<vector<int>>&dp){
        if(r ==0 || r == n) return 1;
        if(r == 1 || r== n-1 ) return n;
        if(dp[n][r] != -1)return dp[n][r];
        return dp[n][r] = ncr(n-1,r,dp) + ncr(n-1,r-1,dp) ;
        
    }
    vector<int> getRow(int rowIndex) {
        vector<int>ans;
        vector<vector<int>>dp(rowIndex+1,vector<int>(rowIndex+1,-1));
        for(int i=0;i<=rowIndex;i++){
            ans.push_back(ncr(rowIndex,i,dp));
        }
        return ans;
    }
};